src/DocumentDbTests/Reading/Linq/invoking_query_with_custom_field_source.cs

```diff
@@ -0,0 +1,263 @@
+                _output.WriteLine("{1} {0}: {2}", parameter.ParameterName, parameter.DbType, parameter.NpgsqlValue);
+        }
+
+        public class MyClass
+        {
+            public string Id { get; set; }
+            public CustomId CustomId { get; set; }
+        }
+
+        public class MyClassNullable
+        {
+            public string Id { get; set; }
+            public CustomId? CustomId { get; set; }
+        }
+
+        public class MyClassArray
+        {
+            public string Id { get; set; }
+            public CustomId[] CustomIds { get; set; }
+        }
+
+        public readonly partial struct CustomId
+        {
+            public CustomId(string value)
+            {
+                Value = value ?? throw new ArgumentNullException(nameof(value));
+            }
+
+            public string Value { get; }
+        }
+
+        public class CustomIdField: FieldBase
+        {
+            public CustomIdField(string dataLocator, Casing casing, MemberInfo[] members): base(dataLocator, "varchar",
+                casing, members)
+            {
+            }
+
+            public override ISqlFragment CreateComparison(string op, ConstantExpression value,
+                Expression memberExpression)
+            {
+                if (value.Value == null)
+                {
+                    return op == "=" ? new IsNullFilter(this) : new IsNotNullFilter(this);
+                }
+
+                var def = new CommandParameter(((CustomId)value.Value).Value);
+                return new ComparisonFilter(this, def, op);
+            }
+
+            public override string SelectorForDuplication(string pgType)
+            {
+                return RawLocator.Replace("d.", "");
+            }
+        }
+
+        public class CustomIdFieldSource: IFieldSource
+        {
+            public bool TryResolve(string dataLocator, StoreOptions options, ISerializer serializer,
+                Type documentType,
+                MemberInfo[] members, out IField? field)
+            {
+                var fieldType = GetRawMemberType(members.Last());
+
+                if (fieldType == null)
+                {
+                    field = null;
+                    return false;
+                }
+
+                if (fieldType == typeof(CustomId))
+                {
+                    field = new CustomIdField(dataLocator, serializer.Casing, members);
+                    return true;
+                }
+
+                if (fieldType.IsNullable() && fieldType.GetGenericArguments()[0] == typeof(CustomId))
+                {
+                    field = new NullableTypeField(new CustomIdField(dataLocator, serializer.Casing, members));
+                    return true;
+                }
+
+                field = null;
+                return false;
+            }
+
+            private static Type? GetRawMemberType(MemberInfo member)
```

Comment:
There's an extension method in either Marten or Baseline to do this already. GetMemberType() I believe.

Reply:
I updated the extension method to make it public